datetime/test.ts

```diff
@@ -64,6 +65,94 @@ Deno.test({
   },
 });
 
+Deno.test("[std/datetime] parse: The date is 2021-12-31", () => {
+  const time = new FakeTime("2021-12-31");
+  try {
+    assertEquals(
+      datetime.parse("01-01", "MM-dd"),
+      new Date(2021, 0, 1),
+    );
+    assertEquals(
+      datetime.parse("02-01", "MM-dd"),
+      new Date(2021, 1, 1),
+    );
+    assertEquals(
+      datetime.parse("03-01", "MM-dd"),
+      new Date(2021, 2, 1),
+    );
+    assertEquals(
+      datetime.parse("04-01", "MM-dd"),
+      new Date(2021, 3, 1),
+    );
+    assertEquals(
+      datetime.parse("05-01", "MM-dd"),
+      new Date(2021, 4, 1),
+    );
+    assertEquals(
+      datetime.parse("06-01", "MM-dd"),
+      new Date(2021, 5, 1),
+    );
+    assertEquals(
+      datetime.parse("07-01", "MM-dd"),
+      new Date(2021, 6, 1),
+    );
+    assertEquals(
+      datetime.parse("08-01", "MM-dd"),
+      new Date(2021, 7, 1),
+    );
+    assertEquals(
+      datetime.parse("09-01", "MM-dd"),
+      new Date(2021, 8, 1),
+    );
+    assertEquals(
+      datetime.parse("10-01", "MM-dd"),
+      new Date(2021, 9, 1),
+    );
+    assertEquals(
+      datetime.parse("11-01", "MM-dd"),
+      new Date(2021, 10, 1),
+    );
+    assertEquals(
+      datetime.parse("12-01", "MM-dd"),
+      new Date(2021, 11, 1),
+    );
+
+    assertEquals(
+      datetime.parse("01", "dd"),
+      new Date(2021, 11, 1),
+    );
+    assertEquals(
+      datetime.parse("15", "dd"),
+      new Date(2021, 11, 15),
+    );
+    assertEquals(
+      datetime.parse("31", "dd"),
+      new Date(2021, 11, 31),
+    );
+
+    assertEquals(
+      datetime.parse("01", "MM"),
+      new Date(2021, 0, 30),
+    );
+
+    assertEquals(
+      datetime.parse("12", "MM"),
+      new Date(2021, 11, 30),
+    );
+
+    assertEquals(
+      datetime.parse("2020", "yyyy"),
+      new Date(2020, 11, 30),
+    );
+    assertEquals(
+      datetime.parse("2022", "yyyy"),
+      new Date(2022, 11, 30),
+    );
```

Comment:
I added these because there were only tests for combination of month and day. I thought it would be good to have additional test coverage for the `dd`, `MM`, and `yyyy` formats.

When writing these tests, I was a bit surprised that the day was changing from 31 to 30. I think that is a bug. If it is and gets fixed, these MM and yyyy test case's expectated dates will need updated to have 31 for the day instead of 30.

I found the cause is a line in `DateTimeFormatter.prototype.partsToDate` from `formatter.ts`. Apparently `date.setHour(0, 0, 0, 0)` will also change the day because the `setHour` method is using the Timezone of the system rather than UTC but the output is always in UTC. I'm in central timezone, so the time gets set to "00:00:00.000 GMT-0600" which ends up changing the date to being "2021-12-30T06:00:00.000Z"

```ts
Deno.test("example", () => {
  const a = new Date("2021-12-31");
  a.setHours(0, 0, 0, 0);
  const b = new Date("2021-12-31");
  assertEquals(a, b);
});
```

```
AssertionError: Values are not equal:

    [Diff] Actual / Expected

-   2021-12-30T06:00:00.000Z
+   2021-12-31T00:00:00.000Z
```

Here is an example showing that behavior in Firefox developer console.

![image](https://user-images.githubusercontent.com/3241653/163307442-65d00ccc-5bbc-4ddf-9710-2ed4f3bfd2dc.png)

Reply:
I had to remove the `MM` and `yyyy` test cases I previously added. The CI runs in UTC time, so the above issue doesn't happen in CI but does for any developers in a different timezone than UTC.